Added paragraph in astropy.io.fits index.rst describing non-standard  FITS

ndl303 opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/master/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/master/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Astropy coding style guidelines can be found here:
https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions
Our testing infrastructure enforces to follow a subset of the PEP8 to be
followed. You can check locally whether your changes have followed these by
running the following command:

tox -e codestyle

-->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

This pull request is to address documentation for handling non-standard FITS files. It adds a short
paragraph quickly reviewing how non-standard FITS files are handled a simple example for users who need to read a non-standard FITS file. It includes a link to the FITS Verification page.

Changes.rst has not been edited (yet).

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #10153

dhomeier commented:
This looks very useful! Maybe a line could be added explaining that the given example would have `io.fits` interpret the non-standard conforming data the best it can, while emitting only a single warning (for each violation, if I recall the `silentfix+warn` setting right)?

saimn commented:
Thanks for the PR. 
I think this needs to be rephrased a bit to highlight a few things: io.fits does not verify inputs, except in a few cases (typically when printing a header in the console), so its the user responsibility to verify the file by calling `.verify` if one needs to know if the file is valid or not. But io.fits will run the verification when writing a file. The idea is to be tolerant with input and strict with output. This is the role of the `output_verify` keyword (which defaults to `exception`) for `fits.open` (added recently), `close`, `flush`, `writeto`.

ndl303 commented:
Thanks for the comments **saimn**.
 
I am slightly confused by your statements as they are contrary to my experience. In my experience, when reading Solar Dynamics Observatory HMI Continuum FITS data, ``io.fits`` threw ``VerifyError`` exceptions when reading _nan_ values_ from CARD values from inside method ``Card._parse_value``.  The only way I and others (e.g. sunpy) have found to successfully read the HMI FITS files is to insert **verify** in the code immediately after the  ``open`` statement; the code then works flawlessly.

My observation is that a call to **verify()** is essential when reading (some?) HMI non-compliant files. I think this is contrary to your statements which suggests (to me) it is only a diagnostic when reading. Please let me know if I have misunderstood your comments. As it stands right now I would not know how to change what I have written in the directions you have indicated. Perhaps we can iterate a couple of times as I agree that we don't want to put (or keep) bad information out there. I would be happy to run a simple test or two if we need clarification of any issues. I am not a FITS expert by any stretch of the imagination so finger trouble at my end is not out of the question.

On another note: I was speaking to the SDO HMI people and they have already started the process of making their FITS files compliant, which is probably a good thing.

saimn commented:
Indeed, it is a bit of a special case, in your case it fails because when accessing `.data` the code parses the header, and it actually parses all keywords *and values*, while it actually does not need to do that. And in other parts of the code, like when using an `ImageHDU` instead of a compressed one, the same error would not have failed. 
In general the idea is to avoid a crash like yours when just reading a file, unless there is a good reason, like a mandatory keyword that is missing or cannot be interpreted. Here the keyword causing the crash is not useful to read the data, so it should not have been read, and you would have noticed the error only when printing the header (with warnings instead of an exception), or when saving the file (with an exception this time). 
Hope it clarifies things a bit, which is not simple, sorry!

ndl303 commented:
Hi @saimn.  Sorry for the long response. I agree that in general you want to avoid a crash (who does not?). But how do I actually pull that off?  I tried fixing the problem cards before reading the data but I still get exceptions and errors when reading the data field.  To illustrate this, I tried three examples:

1.  open the file and load the cards to find out the problems cards. This identifies 4 bad cards.
2.  open the file and fix the 4 problem cards before loading the data. This claims to fix the cards but still fails when I load the data. 
3. open the file , call `hdul.verify` and then load the data. This works flawlessly.

```
hdul=fits.open(filename)
cards = hdul[1].header.cards
WARNING: VerifyWarning: Verification reported errors: [astropy.io.fits.verify]
WARNING: VerifyWarning: Card 'CRDER1' is not FITS standard (invalid value string: 'nan').  Fixed 'CRDER1' card to meet the FITS standard. [astropy.io.fits.verify]
WARNING: VerifyWarning: Note: astropy.io.fits uses zero-based indexing.
 [astropy.io.fits.verify]
WARNING: VerifyWarning: Card 'CRDER2' is not FITS standard (invalid value string: 'nan').  Fixed 'CRDER2' card to meet the FITS standard. [astropy.io.fits.verify]
WARNING: VerifyWarning: Card 'CSYSER1' is not FITS standard (invalid value string: 'nan').  Fixed 'CSYSER1' card to meet the FITS standard. [astropy.io.fits.verify]
WARNING: VerifyWarning: Card 'CSYSER2' is not FITS standard (invalid value string: 'nan').  Fixed 'CSYSER2' card to meet the FITS standard. [astropy.io.fits.verify]
hdul.close()
```
```
hdul=fits.open(filename)
hdul[1].header.cards['CRDER2'].verify('silentfix+warn')
hdul[1].header.cards['CRDER1'].verify('silentfix+warn')
hdul[1].header.cards['CSYSER1'].verify('silentfix+warn')
hdul[1].header.cards['CSYSER2'].verify('silentfix+warn')
data = hdul[1].data
Traceback (most recent call last):
  File "C:\Users\nickl\Anaconda3\envs\altius\lib\site-packages\IPython\core\interactiveshell.py", line 3331, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-31-4b341e9665e2>", line 1, in <module>
    data = hdul[1].data
  File "C:\Users\nickl\Anaconda3\envs\altius\lib\site-packages\astropy\utils\decorators.py", line 758, in __get__
    val = self.fget(obj)
  File "C:\Users\nickl\Anaconda3\envs\altius\lib\site-packages\astropy\io\fits\hdu\compressed.py", line 1363, in data
    data = compression.decompress_hdu(self)
  File "C:\Users\nickl\Anaconda3\envs\altius\lib\site-packages\astropy\utils\decorators.py", line 758, in __get__
    val = self.fget(obj)
  File "C:\Users\nickl\Anaconda3\envs\altius\lib\site-packages\astropy\io\fits\hdu\compressed.py", line 1415, in compressed_data
    compressed_data = super().data
  File "C:\Users\nickl\Anaconda3\envs\altius\lib\site-packages\astropy\utils\decorators.py", line 758, in __get__
    val = self.fget(obj)
  File "C:\Users\nickl\Anaconda3\envs\altius\lib\site-packages\astropy\io\fits\hdu\table.py", line 399, in data
    data = self._get_tbdata()
  File "C:\Users\nickl\Anaconda3\envs\altius\lib\site-packages\astropy\io\fits\hdu\table.py", line 156, in _get_tbdata
    columns = self.columns
  File "C:\Users\nickl\Anaconda3\envs\altius\lib\site-packages\astropy\utils\decorators.py", line 758, in __get__
    val = self.fget(obj)
  File "C:\Users\nickl\Anaconda3\envs\altius\lib\site-packages\astropy\io\fits\hdu\table.py", line 395, in columns
    return self._columns_type(self)
  File "C:\Users\nickl\Anaconda3\envs\altius\lib\site-packages\astropy\io\fits\column.py", line 1405, in __init__
    self._init_from_table(input)
  File "C:\Users\nickl\Anaconda3\envs\altius\lib\site-packages\astropy\io\fits\column.py", line 1470, in _init_from_table
    for keyword, value in hdr.items():
  File "C:\Users\nickl\Anaconda3\envs\altius\lib\site-packages\astropy\io\fits\header.py", line 983, in items
    yield (card.keyword, card.value)
  File "C:\Users\nickl\Anaconda3\envs\altius\lib\site-packages\astropy\io\fits\card.py", line 285, in value
    value = self._value = self._parse_value()
  File "C:\Users\nickl\Anaconda3\envs\altius\lib\site-packages\astropy\io\fits\card.py", line 767, in _parse_value
    ".verify('fix').".format(self.keyword))
astropy.io.fits.verify.VerifyError: Unparsable card (CRDER1), fix it first with .verify('fix').
hdul.close()
```

```
hdul=fits.open(filename)
hdul.verify('silentfix+warn')
data = hdul[1].data

```
I was unable to find any function or method within 10-20 minutes of browsing the documentation that would convert the CompImageHDU  to a ImageHDU so I could not write any code to look at this solution.

It strikes me that, as presented through the Astropy FITS documentation, a novice user must go through `.data` to get to the data they want, which implies they will read all the header fields and values. If this fails, which it did for me, then if no solution is forthcoming after reading the primary documentation, there is a good chance they will either abandon astropy.io.fits as "not working" or too difficult to use or they will post questions to astropy support (which is what I did).  None of these scenarios are particularly desirable for anyone.

I think my intent with this pull request is to provide a short paragraph on the primary FITS page that gives a short practical account of what to do if someone encounters a verification exception when reading a FITS file.  I think my few lines describing the `hdul.verify()` option do a reasonable job of meeting that goal and direct the interested reader to the verification page for more detailed info. 

I would be happy to change the text in the paragraph if you feel it is inaccurate, misleading or insufficient but I have not yet seen or found anything to suggest this.  I can also accept that the pull request is not implemented if the circumstances I have encountered are considered too specific to be of use to the general FITS user. I would personally disagree with that decision but I would accept the review process.

dhomeier commented:
> It strikes me that, as presented through the Astropy FITS documentation, a novice user must go through `.data` to get to the data they want, which implies they will read all the header fields and values. If this fails, which it did for me, then if no solution is forthcoming after reading the primary documentation, there is a good chance they will either abandon astropy.io.fits as "not working" or too difficult to use or they will post questions to astropy support (which is what I did). None of these scenarios are particularly desirable for anyone.

Since `data` is the hdu property containing the actual data, it is quite inevitable to go through it, at least indirectly. But if I understand #10171 correctly, that PR should already fix avoidable crashes on accessing `data`, as long as all _required_ cards for interpreting it have valid values, is that right @saimn?
I still think _this_ PR adds valuable documentation on how to use `.verify` even once #10171 is merged, but then it probably needs to be clarified what cases remain where the data cannot be read with the default settings, but `verify` would resolve them easily – e.g. could `silentfix` also change required, but missing or invalid cards, to a default value?

saimn commented:
Yes, thanks @dhomeier , with #10171 you would be able to open the file and access the data without crashing. Because we need to parse only the *structural* keywords, the ones useful to read the data. The same is already true for image data:
```
In [15]: hdul = fits.open('test.fits')                                                         

In [17]: hdul[1].data                                                                          
Out[17]: array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

In [18]: hdul[1].header                                                                        
Out[18]: WARNING: VerifyWarning: Verification reported errors: [astropy.io.fits.verify]
WARNING: VerifyWarning: Card 'FOO' is not FITS standard (invalid value string: 'NAN.0').  Fixed 'FOO' card to meet the FITS standard. [astropy.io.fits.verify]
WARNING: VerifyWarning: Note: astropy.io.fits uses zero-based indexing.
 [astropy.io.fits.verify]

XTENSION= 'IMAGE   '           / Image extension                                
BITPIX  =                  -64 / array data type                                
NAXIS   =                    1 / number of array dimensions                     
NAXIS1  =                   10                                                  
PCOUNT  =                    0 / number of parameters                           
GCOUNT  =                    1 / number of groups                               
FOO     = 'NAN.0   '                                 
```
By default `io.fits` makes it hard to write such an invalid file, but the idea is that you should be able to read existing files even if they are invalid.

Then I also agree that the documentation addition is useful, I just wanted to explain a bit more the current behavior. And the verification documentation (http://docs.astropy.org/en/latest/io/fits/usage/verification.html) states clearly that *When astropy reads a FITS file which does not conform to FITS standard, it will not raise an error and exit*.

saimn commented on docs/io/fits/index.rst:
```diff
@@ -178,6 +178,22 @@ When writing a file (e.g., with the :func:`writeto` function), compression will
+
+Working with non-standard files
+"""""""""""""""""""""""""""""""
+`astropy.io.fits` adheres to the FITS standard and verifies that all fields within a FITS file conform
+to the standard. It may raise ``VerifyError`` if a field is deemed to be non-compliant. This action can be modified
+using method ``HDUList.verify``. It is typically invoked after opening the file but before reading any headers or data::
```
Suggestion:

When astropy reads a FITS file which does not conform to FITS standard, it will not raise an error and exit. It will try to make the best educated interpretation and only gives up when the offending data is accessed and no unambiguous interpretation can be reached. If the file contains some invalid keyword, it may trigger warnings later when printing the header, or an exception when trying to write the file.
The file can be verified and fixed with `HDUList.verify`, and this can also be controlled when writing the file with the `output_verify` parameter.

ndl303 replied:
I am very happy with this overall direction as it looks like we are converging on a sensible addition to the FITS documentation.

I am confused in the first sentence when when you say "it will not raise an error and exit". This is the exact opposite of my experience: astropy does raise an error and exit when reading  non-compliant FITS files generated by the solar data observatory HMI instrument, (unless you include the "verify" option).  This is the reason why I decided to write the extra paragraph in the FITS documentation, as it took me some time and false attempts to find the `verify` solution, e.g I tried using `output_verify` which is documented in the `open` method but that did not work for me when reading files.

It seems from the previous messages that you are basing your statement _"it will not raise an error and exit"_ upon information on the verification page (http://docs.astropy.org/en/latest/io/fits/usage/verification.html). I believe this sentence is slightly misleading. I think it should say _when astropy reads a FITS file which does not conform to FITS standard, it will not IMMEDIATELY raise an error and exit_. 

The current version of Astropy does eventually raise an error if it cannot decode a value when reading a file. For example, I raise an error near line 766 of card.py in method `_parse_value` when I try to read my FITS file. If I look at this code I see,

```
        m = self._value_NFSC_RE.match(self._split()[1])

        if m is None:
            raise VerifyError("Unparsable card ({}), fix it first with "
                              ".verify('fix').".format(self.keyword))

```
Clearly a `VerifyError` is raised if `m is None` when `self._split()[1]` fails to match the regular expression. 

Am I making some mistake here and just missing the point? I apologize if that is the case.
I am new to making modifications to Astropy and I am still feeling my way through the system.

I think if we can get this issue clarified the pull request will sort itself out quite quickly. I am happy to run some test cases if that would help clarify any questions.

ndl303 replied:
OK. I think I am slowly coming up to speed. Are you suggesting that we have a paragraph in the documentation that is consistent with the changes applied by @dhomeier  with #10171 ? This would then make your suggested paragraph make sense to me.

saimn replied:
#10171 will fix the crash with your file when reading the data. So then the sentence "it will not raise an error and exit" will be true (hopefully, there may still be cases where it could crash, if a required keyword is invalid, but hopefully this should not be so frequent).

saimn commented on docs/io/fits/index.rst:
```diff
@@ -178,6 +178,22 @@ When writing a file (e.g., with the :func:`writeto` function), compression will
+
+    >>> hdul = fits.open(fits_image_filename)
+    >>> hdul.verify('silentfix+warn')
+    >>> data = hdul.data
```
Please use the context manager here, `with fits.open(fits_image_filename) as hdul: ...`. Also for the educational purpose of the example I would prefer `fix` instead of `silentfix` ;).

ndl303 replied:
OK. Sounds good to me.  I was following the format of earlier examples on the documentation page which do not use the "with" statement. I'm ok with 'fix'

saimn commented on docs/io/fits/index.rst:
```diff
@@ -178,6 +178,27 @@ When writing a file (e.g., with the :func:`writeto` function), compression will
+after opening the file but before accessing any headers or data::
+
+    >>> with fits.open(fits_image_filename) as hdul:
+    >>>    hdul.verify('fix')
+    >>>    data = hdul[1].data
```
To fix the doctest:
```suggestion
    ...    hdul.verify('fix')
    ...    data = hdul[1].data
```

saimn commented:
I removed the merge commit and my commit which was conflicting with yours, so this is now good to go. Thanks for the contribution @ndl303 !
